bpf: Add 'host not ready' drop reason if host policy cannot be enforced

ti-mo opened the pull request:
Repurpose unused drop reasons since https://github.com/cilium/cilium/commit/61fb508f2c7320e6cf29f0024f678e5ea22e5d15
("bpf: Rename unused drop defines to DROP_UNUSED*") for signaling the host
endpoint's policy program was attempted to be executed before it was loaded.

bpf_lxc.c contains multiple tail calls into POLICY_CALL_MAP at the HOST_EP_ID
slot. The program in this slot is provided by bpf_host.c. During first agent
startup, there are often multiple Pods pending creation due to no CNI being
available. As soon as the agent's local API becomes available, these outstanding
CNI requests have a chance to be accepted by the API handler.

If one such request is serviced before the host datapath controller has a chance
to grab the compilation lock, the endpoint program will compile and attach first.
If the host firewall is enabled, and this new workload endpoint sends a packet
to the host, the host's ingress policy needs to be enforced. However, because
bpf_host hasn't been loaded yet, this policy program is not yet present in
POLICY_CALL_MAP, resulting in a missed tail call.

One potential solution to this problem would be making sure the host datapath
always attaches before workload endpoints do. There's one problem with this
solution: clustermesh requires data from other clusters in order to correctly
populate the local ipcache, and the ipcache currently needs to be populated for
the host endpoint to finish attaching. It obtains this information through
clustermesh-apiserver, typically deployed onto the local node as a regular Pod.
This means workload endpoints must be able to deploy before the host endpoint.

As a stop gap, tolerate these kinds of drops and assign them a specific meaning,
without letting them spill over into the generic 'missed tail call' bucket. To
stabilize end-to-end tests, we're aiming to enforce zero dropped tail calls in
all CI scenarios, since it leads to packets that mysteriously go missing,
introducing chaos that's impossible to troubleshoot.


```release-note
Add specific drop reason for missing tail calls if the host datapath is not ready yet
```

aanm reviewed (commented):
Don't we also need to wait for the hostdatapath to be initialized during restoration of endpoints?

tommyp1ckles commented on daemon/cmd/state.go:
```diff
@@ -318,39 +318,30 @@ func (d *Daemon) regenerateRestoredEndpoints(state *endpointRestoreState, endpoi
+		// Cap the timeout used to wait for remote cluster synchronization
+		// to avoid blocking the agent startup, as this regeneration is
+		// performed synchronously.
+		endpointsRegenerator.CapTimeoutForSynchronousRegeneration()
```
The function comment for `CapTimeoutForSynchronousRegeneration` still says this is only used when ipsec is enabled.

ti-mo replied:
Hmm, that might explain the clustermesh failures. I'll lift it out of the loop,

tommyp1ckles commented on daemon/cmd/state.go:
```diff
@@ -318,39 +318,30 @@ func (d *Daemon) regenerateRestoredEndpoints(state *endpointRestoreState, endpoi
 
 	for _, ep := range state.restored {
-		if ep.IsHost() && option.Config.EnableIPSec {
+		if ep.IsHost() {
```
Why do we not skip regenerating here again if the sync one happened above?

ti-mo replied:
Sorry, I don't understand the question. The first loop handles the host endpoint, this loop handles all the others.

ti-mo replied:
Removed this code for now, it may be the cause of clustermesh test failures.

tommyp1ckles commented on daemon/cmd/endpoint.go:
```diff
@@ -569,6 +569,12 @@ func putEndpointIDHandler(d *Daemon, params PutEndpointIDParams) (resp middlewar
+	// endpoint programs. bpf_lxc makes the assumption that bpf_host is loaded
+	// and its policy program(s) inserted. Can't live in createEndpoint as that
+	// function is called from tests that don't set up a host datapath.
+	<-d.Datapath().Loader().HostDatapathInitialized()
```
Makes sense to me,  blocking inside the api handler does seem a bit strange to me however.

One possible improvement on this could be to `select {}` on both this and the request context in case the request is cancelled prior to this completing.

```go
select {
    case <-d.Datapath().Loader().HostDatapathInitialized():
    case <-d.apiLimiterSet.Wait(params.HTTPRequest.Context().Done():
        // ...
}
```

ti-mo replied:
The code in question has been removed.

rolinh commented on api/v1/flow/flow.proto:
```diff
@@ -335,7 +335,7 @@ enum DropReason {
     DROP_REASON_UNKNOWN = 0;
     // drop reasons
-    INVALID_SOURCE_MAC = 130;
+    HOST_NOT_READY = 130;
```
Don't reuse a field number as it messes up deserialization. A client with an older definition of the API would report `INVALID_SOURCE_MAC` and not `HOST_NOT_READY`.
Best practice here if this field is unused is to  mark it as deprecated like below and pick the next available field number for `HOST_NOT_READY`.
```suggestion
    // Deprecated: this drop reason is not emitted since Cilium version x.y.z
    HOST_NOT_READY = 130 [deprecated = true];
```

rolinh reviewed (approved):
API changes lgtm.

tommyp1ckles commented on bpf/bpf_lxc.c:
```diff
@@ -1053,11 +1053,13 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx, __u32 *d
 		ctx_store_meta(ctx, CB_FROM_HOST, 0);
 		tail_call_static(ctx, &POLICY_CALL_MAP, HOST_EP_ID);
-		return DROP_MISSED_TAIL_CALL;
+		return DROP_HOST_NOT_READY;
```
Will this case, and the `send_drop_notify` case below always be as a result of the host policy program not being in place?

ti-mo replied:
Yes, if calling into POLICY_CALL_MAP at slot HOST_EP_ID fails, the host program has not been loaded.

tommyp1ckles commented on api/v1/flow/flow.proto:
```diff
@@ -335,15 +335,15 @@ enum DropReason {
     // drop reasons
-    INVALID_SOURCE_MAC = 130;
-    INVALID_DESTINATION_MAC = 131;
+    INVALID_SOURCE_MAC = 130 [deprecated = true];
```
Are these deprecations a consequence of the new drop reason?  If not it might be worth putting this into a separate commit to separate-out & explain the change.

ti-mo replied:
I initially lobbed it in to avoid regenerating the proto twice, easing backports, but with the IGMP drop reasons  added in between we'll need to regenerate when backporting anyway.

ti-mo committed:
ci: conformance-e2e: ignore 'Host datapath not ready' drop reason

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
api: mark deprecated drop reasons as such in flow.proto

These reasons are no longer in use by the datapath. Explicitly mark them
as deprecated in the proto to avoid confusion.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
bpf: Add 'host not ready' drop reason if host policy cannot be enforced

bpf_lxc.c contains multiple tail calls into POLICY_CALL_MAP at the HOST_EP_ID
slot. The program in this slot is provided by bpf_host.c. During first agent
startup, there are often multiple Pods pending creation due to no CNI being
available. As soon as the agent's local API becomes available, these outstanding
CNI requests have a chance to be accepted by the API handler.

If one such request is serviced before the host datapath controller has a chance
to grab the compilation lock, the endpoint program will compile and attach first.
If the host firewall is enabled, and this new workload endpoint sends a packet
to the host, the host's ingress policy needs to be enforced. However, because
bpf_host hasn't been loaded yet, this policy program is not yet present in
POLICY_CALL_MAP, resulting in a missed tail call.

One potential solution to this problem would be making sure the host datapath
always attaches before workload endpoints do. There's one problem with this
solution: clustermesh requires data from other clusters in order to correctly
populate the local ipcache, and the ipcache currently needs to be populated for
the host endpoint to finish attaching. It obtains this information through
clustermesh-apiserver, typically deployed onto the local node as a regular Pod.
This means workload endpoints must be able to deploy before the host endpoint.

As a stop gap, tolerate these kinds of drops and assign them a specific meaning,
without letting them spill over into the generic 'missed tail call' bucket. To
stabilize end-to-end tests, we're aiming to enforce zero dropped tail calls in
all CI scenarios, since it leads to packets that mysteriously go missing,
introducing chaos that's impossible to troubleshoot.

Signed-off-by: Timo Beckers <timo@isovalent.com>

brlbil reviewed (approved):
CI part LGTM.
